### PR TITLE
Add 3DPoolingLayer

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/438
Add 3DPoolingLayer.

* [x] CPU implementation. 
* [x] GPU implementation.
* [x] Add unit test.
* [x] Add maxPoolIdx
* [x] Add config parser for pooling3D

### CODE REVIEW

**Comment by chengduoZH** [comment]
Start from "Add Pool3DLayer"

**Comment by qingqing01** [comment]
Please remove 3DConv/3D-DeConv in this PR at first.

File: paddle/cuda/include/hl_cnn.h [hunk 1]
```diff
@@ -174,6 +174,202 @@ extern void hl_avgpool_backward(const int frameCnt,
...
+ * @param[in]   strideD     pooling stride depth.
+ * @param[in]   strideH     pooling stride height.
+ * @param[in]   strideW     pooling stride width.
+ * @param[in]   paddingD    padding depth.
+ * @param[in]   paddingH    padding height.
+ * @param[in]   paddingW    padding width.
+ * @param[out]  tgtData     output data.
+ * @param[in]   tgtStride   stride between output data samples.
+ *
+ */
```
Comment (qingqing01) [inline paddle/cuda/include/hl_cnn.h]: remove comments, since the following argument names are clear.

File: paddle/gserver/layers/Pool3DLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  size_t layerSize = 0;
+  //  imgSizeD_ = inputLayers_[0]->getOutput().getFrameDepth();
+  //  imgSizeH_ = inputLayers_[0]->getOutput().getFrameHeight();
+  //  imgSizeW_ = inputLayers_[0]->getOutput().getFrameWidth();
+  if (imgSizeH_ == 0) {
+    //    imgSizeH_ = imgSizeY_;
+  }
+  if (imgSizeW_ == 0) {
+    //    imgSizeW_ = imgSize_;
+  }
```
Comment (qingqing01) [inline paddle/gserver/layers/Pool3DLayer.cpp]: If these codes are out of use, please remove them

File: paddle/math/Matrix.cpp [hunk 3]
```diff
@@ -1190,6 +1190,224 @@ void GpuMatrix::avgPoolBackward(Matrix& outGrad,
...
+                                 size_t strideD,
+                                 size_t strideH,
+                                 size_t strideW,
+                                 size_t outputD,
+                                 size_t outputH,
+                                 size_t outputW,
+                                 size_t paddingD,
+                                 size_t paddingH,
+                                 size_t paddingW) {
+  CHECK(inputMat.useGpu_ == true) << "Matrix type are not equal";
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]:
```
CHECK(inputMat.useGpu_) << "Matrix type is not correct";
```

File: paddle/math/Matrix.cpp [hunk 4]
```diff
@@ -1190,6 +1190,224 @@ void GpuMatrix::avgPoolBackward(Matrix& outGrad,
...
+                                  size_t outputD,
+                                  size_t outputH,
+                                  size_t outputW,
+                                  real scaleTargets,
+                                  real scaleOutput,
+                                  size_t paddingD,
+                                  size_t paddingH,
+                                  size_t paddingW) {
+  CHECK(inputMat.useGpu_ == true && outGrad.useGpu_ == true &&
+        outV.useGpu_ == true)
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]:
```
CHECK(inputMat.useGpu_  && outGrad.useGpu_  && outV.useGpu_)
```

File: paddle/math/Matrix.cpp [hunk 5]
```diff
@@ -1190,6 +1190,224 @@ void GpuMatrix::avgPoolBackward(Matrix& outGrad,
...
+                                 size_t strideD,
+                                 size_t strideH,
+                                 size_t strideW,
+                                 size_t outputD,
+                                 size_t outputH,
+                                 size_t outputW,
+                                 size_t paddingD,
+                                 size_t paddingH,
+                                 size_t paddingW) {
+  CHECK(inputMat.useGpu_ == true) << "Matrix type are not equal";
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]: same as above.

File: paddle/math/Matrix.cpp [hunk 6]
```diff
@@ -1190,6 +1190,224 @@ void GpuMatrix::avgPoolBackward(Matrix& outGrad,
...
+                       sizeZ,
+                       sizeY,
+                       sizeX,
+                       strideD,
+                       strideH,
+                       strideW,
+                       paddingD,
+                       paddingH,
+                       paddingW,
+                       data_,
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]: data_ -> getData()

File: paddle/math/Matrix.cpp [hunk 7]
```diff
@@ -1190,6 +1190,224 @@ void GpuMatrix::avgPoolBackward(Matrix& outGrad,
...
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]: data_ - > getData()

File: paddle/math/Matrix.cpp [hunk 8]
```diff
@@ -1190,6 +1190,224 @@ void GpuMatrix::avgPoolBackward(Matrix& outGrad,
...
+                        sizeX,
+                        strideD,
+                        strideH,
+                        strideW,
+                        paddingD,
+                        paddingH,
+                        paddingW,
+                        scaleTargets,
+                        scaleOutput,
+                        data_,
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]: same as above

File: paddle/gserver/layers/Pool3DLayer.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  outputD_ = outputSize(imgSizeD_,
+                        sizeZ_,
+                        paddingD_,
+                        strideD_,
+                        /* caffeMode */ false);
```
Comment (qingqing01) [inline paddle/gserver/layers/Pool3DLayer.cpp]: 这里是为了上取整，去掉`caffeMode`, 放在这里容易让人误解~

File: paddle/gserver/layers/Pool3DLayer.cpp [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+                                 sizeY_,
+                                 sizeZ_,
+                                 strideD_,
+                                 strideH_,
+                                 strideW_,
+                                 outputD_,
+                                 outputH_,
+                                 outputW_,
+                                 1,
+                                 1,
```
Comment (qingqing01) [inline paddle/gserver/layers/Pool3DLayer.cpp]: 1.0

File: paddle/math/Matrix.h [hunk 11]
```diff
@@ -1506,6 +1670,82 @@ class CpuMatrix : public Matrix {
                        real scaleOutput,
                        size_t paddingH,
                        size_t paddingW);
+  //////////////////////
```
Comment (qingqing01) [inline paddle/math/Matrix.h]: remove this line.

File: paddle/math/Matrix.h [hunk 12]
```diff
@@ -1348,6 +1435,83 @@ class GpuMatrix : public Matrix {
...
+                         size_t strideH,
+                         size_t strideW,
+                         size_t outputD,
+                         size_t outputH,
+                         size_t outputW,
+                         real scaleTargets,
+                         real scaleOutput,
+                         size_t paddingD,
+                         size_t paddingH,
+                         size_t paddingW);
```
Comment (qingqing01) [inline paddle/math/Matrix.h]: 虽然无关紧要，但是这些参数的顺序稍乱， 尤其把`real scaleTargets`, `real scaleOutput`莫名的插在中间~

File: paddle/math/tests/test_matrixCompare.cpp [hunk 13]
```diff
@@ -1203,4 +1204,207 @@ TEST(Matrix, warpCTC) {
...
+  int outH = outputSize(height, filterY, padZ, strideY, true);
+  int outW = outputSize(width, filterX, padZ, strideX, true);
+
+  int colBufWidth = outD * outH * outW;
+  MatrixPtr cpuOutput = std::make_shared<CpuMatrix>(1, channel * colBufWidth);
+  MatrixPtr gpuOutput = std::make_shared<GpuMatrix>(1, channel * colBufWidth);
+
+  cpuImage->randomizeUniform();
+  gpuImage->copyFrom(*cpuImage);
+  // std::cout << "test maxPool3DForward...\n";
```
Comment (qingqing01) [inline paddle/math/tests/test_matrixCompare.cpp]: remove this line.

File: paddle/math/tests/test_matrixCompare.cpp [hunk 14]
```diff
@@ -1203,4 +1204,207 @@ TEST(Matrix, warpCTC) {
...
+                              padZ,
+                              padY,
+                              padX);
+  TensorCheckErr(*cpuImage, *gpuImage);
+}
+
+TEST(Matrix, Pool3D) {
+  for (auto depth : {9, 16, 64, 128}) {
+    for (auto height : {9, 11, 128, 256}) {
+      for (auto width : {9, 32, 128}) {
```
Comment (qingqing01) [inline paddle/math/tests/test_matrixCompare.cpp]: 减少测试case~

File: paddle/gserver/layers/Pool3DLayer.cpp (hunk 10 above)
Comment (chengduoZH) [inline paddle/gserver/layers/Pool3DLayer.cpp]: Done

**Pull request merged**